src/components/DeviceInteraction/DeviceInteractionStep.js

```diff
@@ -0,0 +1,221 @@
+    const stillActivated = isActive && prevProps.isActive
+    const didResetError = !error && !!prevProps.error
+
+    if (didActivated && status !== 'running') {
+      this.run()
+    }
+
+    if (didResetError && stillActivated) {
+      this.run()
+    }
+
+    if (didDeactivated && status === 'running') {
+      this.cancel()
+    }
+  }
+
+  componentWillUnmount() {
+    if (this._unsubscribe) {
+      this._unsubscribe()
+    }
+    this._unmounted = true
+  }
+
+  __IS_MOUNTED__THX_FOR_REMOVING_COMPONENTWILLMOUNT__ = false
+  _unsubscribe = null
+  _unmounted = false
+
+  handleSuccess = (res: any) => {
+    const { onSuccess, step, isError } = this.props
+    if (isError) return
+    this.safeSetState({ status: 'idle' })
+    onSuccess(res, step)
+  }
+
+  handleFail = (e: Error) => {
+    const { onFail, step } = this.props
+    this.safeSetState({ status: 'idle' })
+    onFail(e, step)
+  }
+
+  safeSetState = (...args: any) => {
+    if (this._unmounted) return
+    this.setState(...args)
+  }
+
+  run = async () => {
+    const { step, data } = this.props
+
+    if (this.__IS_MOUNTED__THX_FOR_REMOVING_COMPONENTWILLMOUNT__) {
+      this.safeSetState({ status: 'running' })
+    }
+
+    if (!step.run) {
+      return
+    }
+
+    try {
+      const d1 = Date.now()
+
+      // $FlowFixMe JUST TESTED THE `run` 6 LINES BEFORE!!!
+      const res = (await step.run(data)) || {}
+
+      if (step.minMs) {
+        const d2 = Date.now()
+        // $FlowFixMe SAME THING, JUST TESTED THE MINMS KEY, BUT EH
+        if (d2 - d1 < step.minMs) {
+          // $FlowFixMe nice type checking
+          await delay(step.minMs - (d2 - d1))
+        }
+      }
+      if (res.promise) {
+        this._unsubscribe = res.unsubscribe
+        const realRes = await res.promise
+        this.handleSuccess(realRes)
+      } else {
+        this.handleSuccess(res)
+      }
+    } catch (e) {
+      this.handleFail(e)
+    }
+  }
+
+  cancel = () => this.safeSetState({ status: 'idle' })
+
+  render() {
+    const {
+      isFirst,
+      isLast,
+      isActive,
+      isPrecedentActive,
+      isSuccess,
+      isError,
+      isPassed,
+      step,
+      error,
+      onRetry,
+      data,
+    } = this.props
+
+    const { status } = this.state
+
+    const title = typeof step.title === 'function' ? step.title(data) : step.title || '[UNTITLED]'
+
+    return (
+      <DeviceInteractionStepContainer
+        isFirst={isFirst}
+        isLast={isLast}
+        isSuccess={isSuccess}
+        isActive={isActive}
+        isPrecedentActive={isPrecedentActive}
+        isError={isError}
+      >
+        <IconContainer>{step.icon}</IconContainer>
+        <Box py={4} pr={4} justify="center" grow shrink>
+          <Box color={isActive && !isSuccess ? 'dark' : ''} ff="Open Sans|SemiBold">
+            {title}
+          </Box>
+          {step.desc && step.desc}
+          {step.render && (
+            <Box>
+              {step.render(
```

Comment:
it could be a component?

```
<Step.Render
  onSuccess={this.handleSuccess}
  onFail={this.handleFail}
  onRetry={onRetry}
  data={data}
/>
```